### ISSUE TITLE
Returning BigQuery LoadJob causes pickling issue when running flow via agent

### ISSUE THREAD

**Issue by fwinokur** [PrefectHQ/prefect#3657]
## Description
When running a flow via an agent that calls `BigQueryLoadFile`, the returned object (of type google.cloud.bigquery.job.LoadJob) causes a `_pickle.PicklingError` despite a successful run of this task.
```
[2020-11-12 15:08:48-0500] INFO - prefect.CloudTaskRunner | Task 'BigQueryLoadFile': Starting task run...
[2020-11-12 15:08:49-0500] INFO - prefect.CloudTaskRunner | Task 'BigQueryLoadFile': Finished task run for task with final state: 'Mapped'
[2020-11-12 15:08:49-0500] INFO - prefect.CloudTaskRunner | Task 'BigQueryLoadFile[0]': Starting task run...
[2020-11-12 15:09:10-0500] ERROR - prefect.CloudTaskRunner | Unexpected error: PicklingError('Pickling client objects is explicitly not supported.\nClients have non-trivial state that is local and unpickleable.')
Traceback (most recent call last):
  File "/path/to/virtualenv/lib/python3.8/site-packages/prefect/engine/runner.py", line 48, in inner
    new_state = method(self, state, *args, **kwargs)
  File "/path/to/virtualenv/lib/python3.8/site-packages/prefect/engine/task_runner.py", line 891, in get_task_run_state
    result = self.result.write(value, **formatting_kwargs)
  File "/path/to/virtualenv/lib/python3.8/site-packages/prefect/engine/results/gcs_result.py", line 75, in write
    binary_data = new.serializer.serialize(new.value)
  File "/path/to/virtualenv/lib/python3.8/site-packages/prefect/engine/serializers.py", line 70, in serialize
    return cloudpickle.dumps(value)
  File "/path/to/virtualenv/lib/python3.8/site-packages/cloudpickle/cloudpickle_fast.py", line 73, in dumps
    cp.dump(obj)
  File "/path/to/virtualenv/lib/python3.8/site-packages/cloudpickle/cloudpickle_fast.py", line 563, in dump
    return Pickler.dump(self, obj)
  File "/path/to/virtualenv/lib/python3.8/site-packages/google/cloud/client.py", line 163, in __getstate__
    raise PicklingError(
_pickle.PicklingError: Pickling client objects is explicitly not supported.
Clients have non-trivial state that is local and unpickleable.
```

I believe the exception is related to #2178 as the LoadJob object includes the BIgQuery client.  Sadly, this bug is the result of a change I made to solve #3086.

The issue does not occur when running the flow directly from the command line.  I can reproduce this issue by registering my flow, scheduling a run, then starting an agent on my local machine.  My thought is the pickling behavior is different between these two scenarios.

I'm not certain the best approach to resolving this issue.  Passing BigQuery objects directly now sounds like a bad idea, but the information contained within is super useful.  

There are [too many attributes](https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.job.LoadJob.html) to try and account for everything, but when looking at BigQuery logs, they do return a subset of LoadJob attributes that may be a good place from which to draw inspiration (jobName, jobStatistics and jobStatus seem the most relevant for a completed flow).  

```
jobCompletedEvent: {
    eventName: "query_job_completed"     
    job: {
     jobConfiguration: {
      query: {
       createDisposition: "CREATE_IF_NEEDED"        
       defaultDataset: {
       }
       destinationTable: {
        datasetId: "<redacted>"         
        projectId: "<redacted>"         
        tableId: "<redacted>"         
       }
       query: "SELECT * FROM `<redacted>`"        
       queryPriority: "QUERY_INTERACTIVE"        
       statementType: "SELECT"        
       writeDisposition: "WRITE_TRUNCATE"        
      }
     }
     jobName: {
      jobId: "bquxjob_20424f_175bde70639"       
      location: "US"       
      projectId: "<redacted>"       
     }
     jobStatistics: {
      billingTier: 1       
      createTime: "2020-11-12T19:20:39.133Z"       
      endTime: "2020-11-12T19:20:41.064Z"       
      queryOutputRowCount: "203718"       
      referencedTables: [
       0: {
        datasetId: "<redacted>"         
        projectId: "<redacted>"         
        tableId: "<redacted>"         
       }
      ]
      startTime: "2020-11-12T19:20:39.243Z"       
      totalBilledBytes: "56623104"       
      totalProcessedBytes: "55592170"       
      totalSlotMs: "1554"       
      totalTablesProcessed: 1       
     }
     jobStatus: {
      error: {
      }
      state: "DONE"       
     }
    }
   }
  }
```
I feel obligated to fix this since it is my doing; however, I would love some guidance from the core team to understand what a reasonable approach is here.

## Expected Behavior
My hope would be to still access some of LoadJob information post-BigQuery load from within Prefect.

## Reproduction

Code
```
bq_load_file = gcp.BigQueryLoadFile()
bq_load_file_load_jobs = bq_load_file.map(
        file=gzipped_transformed_files,
        dataset_id=unmapped(dataset_id),
        table=unmapped(table_name),
        project=unmapped(gcp_project),
        schema=unmapped(schema),
        create_disposition=unmapped(job.CreateDisposition.CREATE_NEVER),
        source_format=unmapped(job.SourceFormat.NEWLINE_DELIMITED_JSON),
        write_disposition=unmapped(job.WriteDisposition.WRITE_APPEND),
    )
```
Log
```
[2020-11-12 15:08:48-0500] INFO - prefect.CloudTaskRunner | Task 'BigQueryLoadFile': Starting task run...
[2020-11-12 15:08:49-0500] INFO - prefect.CloudTaskRunner | Task 'BigQueryLoadFile': Finished task run for task with final state: 'Mapped'
[2020-11-12 15:08:49-0500] INFO - prefect.CloudTaskRunner | Task 'BigQueryLoadFile[0]': Starting task run...
[2020-11-12 15:09:10-0500] ERROR - prefect.CloudTaskRunner | Unexpected error: PicklingError('Pickling client objects is explicitly not supported.\nClients have non-trivial state that is local and unpickleable.')
Traceback (most recent call last):
  File "/path/to/virtualenv/lib/python3.8/site-packages/prefect/engine/runner.py", line 48, in inner
    new_state = method(self, state, *args, **kwargs)
  File "/path/to/virtualenv/lib/python3.8/site-packages/prefect/engine/task_runner.py", line 891, in get_task_run_state
    result = self.result.write(value, **formatting_kwargs)
  File "/path/to/virtualenv/lib/python3.8/site-packages/prefect/engine/results/gcs_result.py", line 75, in write
    binary_data = new.serializer.serialize(new.value)
  File "/path/to/virtualenv/lib/python3.8/site-packages/prefect/engine/serializers.py", line 70, in serialize
    return cloudpickle.dumps(value)
  File "/path/to/virtualenv/lib/python3.8/site-packages/cloudpickle/cloudpickle_fast.py", line 73, in dumps
    cp.dump(obj)
  File "/path/to/virtualenv/lib/python3.8/site-packages/cloudpickle/cloudpickle_fast.py", line 563, in dump
    return Pickler.dump(self, obj)
  File "/path/to/virtualenv/lib/python3.8/site-packages/google/cloud/client.py", line 163, in __getstate__
    raise PicklingError(
_pickle.PicklingError: Pickling client objects is explicitly not supported.
Clients have non-trivial state that is local and unpickleable.
```

## Environment
`prefect diagnostics` output:
```
{
  "config_overrides": {},
  "env_vars": [
    "PREFECT__CLOUD__AGENT__AUTH_TOKEN",
    "PREFECT__CONTEXT__SECRETS__SLACK_WEBHOOK_URL",
    "PREFECT__CLOUD__AUTH_TOKEN",
    "PREFECT__CONTEXT__SECRETS__GCP_CREDENTIALS"
  ],
  "system_information": {
    "platform": "macOS-10.15.7-x86_64-i386-64bit",
    "prefect_backend": "cloud",
    "prefect_version": "0.13.15",
    "python_version": "3.8.6"
  }
}
```

**Comment by cicdw** [comment]
Hi @fwinokur - the difference between your local run and your agent run is that when running via an Agent, checkpointing is turned on by default which persists task outputs (using `cloudpickle` by default).

In your exact scenario you can set `checkpoint=False` on the `BigQueryLoadFile` task to work around this.  

That being said, as a general rule we require task outputs to be pickleable so that the Task Library is dask-compatible so we'll need to fix this somehow.

Does calling `to_api_repr()` on the object prior to returning it remove the client reference?  If so, maybe that's the best route forward here.

**Comment by fwinokur** [comment]
Ahh, thank you @cicdw, this is super helpful.  That's a great workaround for the interim, and I agree this should be fixed.

to_api_repr() returns the `jobReference` and `configuration` properties which contain information about the job that was created.  I'm trying not to be biased about my specific use-case, but the caller of BigQueryLoadFile would probably know these things in advance.  I think what could potentially be more relevant is the [JobStatistics](https://cloud.google.com/bigquery/docs/reference/rest/v2/Job#JobStatistics) property.  It seems like the python client breaks a lot of these out as individual properties on the LoadJob object (many which seemingly did not make their way into the client documentation) so re-aggregating those to return them doesn't strike me as an elegant approach either.

Would this be a case where there's no good solution in the absence of an upstream fix?  There is an ability to [add callback function](https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.job.LoadJob.html#google.cloud.bigquery.job.LoadJob.add_done_callback), but I wouldn't feel comfortable making a recommendation there.

**Comment by bnaul** [comment]
@fwinokur You're right that there are a lot of `LoadJob` fields to account for, but my guess is that only a couple are non-pickleable (probably GRPC clients). Maybe you can just explicitly set those to `None` before returning the object at the end of the task? It does introduce a bit of a whack-a-mole dynamic as they could rename/add fields in subsequent versions....but at least in my experience those classes are pretty stable so it might not be too bad of a workaround?

The way I've hunted down problematic attributes like this is in IPython:
```
cloudpickle.dumps(load_job)
%debug  # print out the specific attribute that's breaking
load_job.attr = None
cloudpickle.dumps(load_job)
...
```

Best of luck 🔨 :shipit:

**Comment by fwinokur** [comment]
Apologies for taking so long to respond here, and a big thanks to everyone who has already jumped in to comment or resolve!

**Comment by jcrist** [comment]
Fixed by #3724. Closing.
